### PR TITLE
CQ shared store write optimisations

### PR DESCRIPTION
**Pull request by lhoguin**
Another round of CQ shared message store optimisations and refactors, this time around writes.

The first commit is more a refactor than anything, although no longer calling fsync does help in some scenarios. The second commit should provide a noticeable boost in performance when there are no consumers.

@mkuratczyk I have renamed the branch.

### CODE REVIEW

File: deps/rabbit/src/rabbit_msg_store.erl [hunk 1]
```diff
@@ -7,15 +7,14 @@
 
 -module(rabbit_msg_store).
 
--behaviour(gen_server2).
+-behaviour(gen_server).
```
Comment (lhoguin) [inline deps/rabbit/src/rabbit_msg_store.erl:10]: Oops, revert this.

**Comment by michaelklishin** [comment]
The Dialyzer failure is consistent, I can reproduce it locally:

```
rabbit_msg_store.erl:1009:5: Expression produces a value of type 
          'ok' | {'error', atom()}, but this value is unmatched
```

**Comment by lhoguin** [comment]
This is ready for review/merge.

**Comment by mkuratczyk** [comment]
Full test results:
https://snapshots.raintank.io/dashboard/snapshot/oo6eAYwUI2exFVFxdt6mTQIsb7AT2hrt?orgId=2

The biggest difference is in confirm latency, when using `-c 1` (here for 5kb messages, but similar results for all tested):
Screenshot 2023-06-20 at 19 58 06

And when publishing to a long queue (no consumers):
Screenshot 2023-06-20 at 20 02 12

(the sudden drop in throughput is when the queues reach their `max-length` - dropping messages is currently quite expensive)

**Review by mkuratczyk — APPROVED** [review]

**Comment by michaelklishin** [comment]
So, a 40% drop in publisher confirm latency on one workload, and a 60% drop on another. Not bad.

**Pull request merged**